zone.go

```diff
@@ -109,6 +120,133 @@ func (z *Zone) IsInRootZone() bool {
 	return z.IsTLD() && z.IsDelegated()
 }
 
+// Checks to see if a domain is valid according to character set restriction
+// on the zone.
+// Input must be normalized by the client (lowercase, ASCII-encoded).
+func (z *Zone) IsValidDomain(domain string) bool {
+	if !z.isSubdomain(domain) {
+		return false
+	}
+	if len(z.CodePoints) == 0 {
+		return true
+	}
+	// Don't check the suffix since we've already done that
+	return stringsInCodePoints(z.unicodeLabels(domain), z.CodePoints)
+}
+
+// Return the IDN table language the domain matches. Returns an empty string
+// if no IDN tables are defined for the zone or if the domain is not an IDN.
+// Input must be normalized by the client (lowercase, ASCII-encoded).
+func (z *Zone) IDNTable(domain string) (string, error) {
+	if !z.isSubdomain(domain) {
+		return "", errors.New("domain is not a member of the zone")
+	}
+
+	// A blank IDN table language indicates no IDN tables are present
+	if len(z.CodePoints) == 0 {
+		return "", nil
+	}
+
+	// Don't check the suffix since we've already done that, and make sure
+	// we are working with the unicode form, since the ASCII form will never
+	// match an IDN table
+	labels := z.unicodeLabels(domain)
+
+	// Ensure it is an IDN
+	if stringsInCodePoints(labels, asciiCodePoints) {
+		return "", nil
+	}
+
+	if !stringsInCodePoints(labels, z.CodePoints) {
+		return "", errors.New("domain is not a valid member of the zone")
+	}
+
+	for lang, points := range z.IDNTables {
+		if stringsInCodePoints(labels, points) {
+			return lang, nil
+		}
+	}
+
+	return "", errors.New("domain is not a valid IDN of the zone")
+}
+
+func (z *Zone) isSubdomain(domain string) bool {
+	if len(domain) < len(z.Domain)+2 {
+		return false
+	}
+	if domain[len(domain)-len(z.Domain)-1] != '.' {
+		return false
+	}
+	return strings.HasSuffix(domain, z.Domain)
+}
+
+func (z *Zone) unicodeLabels(domain string) []string {
+	prefix, _ := idna.ToUnicode(domain[:len(domain)-len(z.Domain)-1])
+	return strings.Split(prefix, ".")
+}
+
+var asciiCodePoints = []rune{'-', '-', '0', '9', 'a', 'z'}
+
+func stringsInCodePoints(chunks []string, points []rune) bool {
+	for _, c := range chunks {
+		if !stringInCodePoints(c, points) {
+			return false
+		}
+	}
+	return true
+}
+
+func stringInCodePoints(domain string, points []rune) bool {
+	for _, c := range domain {
```

Comment:
At the expense of scanning the string twice, can we find the min/max `rune` in the `domain` string and test if both are `runeInCodePoints`?